### DISCUSSION TITLE
Expensive string parse in Lua resulting in diff consuming too much memory

### DISCUSSION THREAD

**Discussion by Hugobsb** [NeogitOrg/neogit#1426] [General]
The Neogit diff just freezed my Linux here in a way that I had to press the power button to shutdown _(I don't have the SysRq keys enabled to try a kernel shutdown, maybe I should try it later. It can also be Linux misconfiguration)_.
![image](https://github.com/user-attachments/assets/e4eaf773-011c-4e73-a506-4818936b15e7)

I also tried it in my Mac M1 and it freezed the NeoVim process, I had to manually kill it.

That said, I would like to know if someone can elaborate on my discussion here about it.

In my case I have an edge case example of a really large `package-lock.json` file changes.

See the output of `git diff --stat`:
![image](https://github.com/user-attachments/assets/5cc76b3e-239c-4670-9cb0-aa3e34044d27)

Unfortunately, I can't share the _package-lock.json_ file here, but I think anyone would be able to reproduce the issue by creating a large file diff as such.

**Notes:** I didn't open an issue because I would like to discuss first.

- - - -
Edit: I updated the title to be more precise on what was happening.

**Comment by Hugobsb** [comment]
I mean, I can simply use `git diff` instead, it's not about it. The minimum I achieve here is to not accidentally freeze my entire computer by hitting `Tab` by accident in a file that is too much large for Neogit to diff.

**Reply by CKolkey** [reply]
Hey! This is a fun one - in fact, what got me "into" this project was being annoyed at diffs being slow. Anyways. The issue isn't the diff algo, but the number of lines needing to be parsed. I've run into this a lot in lua - it's just expensive to make/parse that many string objects.

That said, we can try to identify bottlenecks. If you can, could you get a profile of the issue?

```lua
-- Start Profiler
:lua require("plenary.profile").start("profile.log", { flame = true })

-- Stop Profiler
:lua require("plenary.profile").stop()
```

edit: You know what, this is stupid easy to reproduce, so don't worry about the profile :)

**Reply by CKolkey** [reply]
[profile.log](https://github.com/user-attachments/files/16338826/profile.log)
(You can use https://www.speedscope.app/ to view it)

<img width="2056" alt="Screenshot 2024-07-22 at 21 18 41" src="https://github.com/user-attachments/assets/cd13e81a-2dcf-415a-8150-ffb0e5f764b0">

I'd say this is pretty clear-cut

**Reply by CKolkey** [reply]
https://github.com/NeogitOrg/neogit/pull/1427 :)

**Reply by Hugobsb** [reply]
Hey, @CKolkey! I'm sorry for the delay, thank you very much for the responses on this.

It seems to be comically fast now. I tested it again with another package-lock.json as below (I lost the original one :c)
<img width="452" alt="image" src="https://github.com/user-attachments/assets/b637dcca-4b18-4c3a-803e-e6955ca8000c">

Here is the [profile.log](https://github.com/user-attachments/files/16363976/profile.log). I am sending a screenshot to make the visualization easier:
![image](https://github.com/user-attachments/assets/e5e0b20f-6432-4c89-b84f-7e7fda53946d)

**Reply by CKolkey** [reply]
No worries, mate. I was a bit surprised to hear about the regression, since I had tested this with a 100k change a while back, and that loaded fine. The culprit was a classic lua fail: string concatenation via `..` in a tight loop. Every single time triggers a rehashing, which gets very expensive... At least it was a simple fix :)

The `__index` call in your profile is the actual loading/parsing of the diff, so at least now it's spending time where I would expect it to be

**Reply by Hugobsb** [reply]
Really nice. Quite interesting though, I will definitely search more about it. I'm marking the discussion as resolved :D

**Reply by CKolkey** [reply]
I recommend this: https://www.lua.org/gems/sample.pdf - I think I've read it every 6 months or so for the last two years haha
